src/gather.jl

```diff
@@ -0,0 +1,49 @@
+export gather, gather_indices
+
+"""
+    gather!(dst, src, idx)
+
+Reverse operation of scatter, which gather data in `src` to destination according to `idx`.
+For each index `k` in `idx`, assign values to `dst` according to
+
+    dst[k...] = src[idx[k]...]
+
+# Arguments
+- `dst`: the destination where data would be assigned to.
+- `src`: the source data to be assigned.
+- `idx`: the mapping for assignment from source to destination.
+"""
+function gather!(dst::AbstractArray{T,N}, src::AbstractArray{T}, idx::AbstractArray{<:IntOrTuple,N}) where {T,N}
+    @assert size(dst) == size(idx) "dst and idx must have the same size."
+    for k = CartesianIndices(idx)
+        @inbounds dst[k] = src[idx[k]]
+    end
+    dst
+end
+
+"""
+    gather(src, idx)
+
+Reverse operation of scatter, which gather data in `src` to destination according to `idx`.
+For each index `k` in `idx`, assign values to `dst` according to
+
+    dst[k...] = src[idx[k]...]
+
+where destination `dst` is created according to `idx`.
+
+# Arguments
+- `src`: the source data to be assigned.
+- `idx`: the mapping for assignment from source to destination.
+"""
+function gather(src::AbstractArray{T}, idx::AbstractArray{<:IntOrTuple}) where {T}
+    dst = Array{T}(undef, size(idx)...)
```

Comment:
I think this should be
```suggestion
    dst = similar(src, axes(idx)...)
```
which will allow for both CuArrays and OffsetArrays. 

It might be worth changing all other `size` to `axes` too, no reason not to allow offsets when it's easy.